### ISSUE TITLE
Generated wrappers include path to SWIG installation in macro comments

### ISSUE THREAD

**Issue by robinst** [swig/swig#202]
In [taglib-ruby](https://github.com/robinst/taglib-ruby/), we commit the wrappers generated by SWIG to the repository, because we don't want to require our users to have SWIG installed.

What's not so nice is that the generated wrappers include some absolute paths to the SWIG installation, which can be different depending on the platform of the developer, e.g. see the changes [here](https://github.com/robinst/taglib-ruby/pull/61/files#diff-1). This causes the changes to be bigger than necessary.

I had a quick look at the swig source, but couldn't figure out if the paths actually need to be in absolute form there, or if they could be changed to be just installation-relative, e.g. `ruby/rubyprimtypes.swg` instead of `/usr/local/share/swig/2.0.12/ruby/rubyprimtypes.swg`. What do you think?

(Another option would be for us to post-process the output of SWIG, but I'd rather not do that if there's a better way.)

**Comment by ojwb** [comment]
I believe the paths encoded inside `@SWIG:/path/to/file,1234,MACRO@` are used to report error locations for macros, so just making them relative would make diagnostics more confusing.

However, when writing this out in `Source/Preprocessor/cpp.c` you could check if the file is under `SwigLib` (see `Source/Modules/main.cxx`) and if so encode that this path is relative to `SwigLib` and then the relative path, then where the path is extracted again insert `SwigLib` again to give the full path back (`Scanner_locator()` in `Source/Swig/scanner.c`).

So in the case above, the marker might be `/*@SWIG~ruby/rubyprimtypes.swg,19,%ruby_aux_method@*/` instead of `/*@SWIG:/usr/local/share/swig/2.0.12/ruby/rubyprimtypes.swg,19,%ruby_aux_method@*/` (the `~` vs `:` indicating if this is path relative to `SwigLib` or just a relative path).

**Comment by robinst** [comment]
Any updates on this? This would really help for reproducible builds.

**Comment by ojwb** [comment]
I don't think anyone's shown interest in working on it.

If you or somebody else reading is motivated enough to want to work on a patch I sketched out a possible solution above, but maybe better would be to work out how these get to the final generated output and stop that happening.  I believe they're added by SWIG's prepocessor as it expands macros so that SWIG's lexer can track the origin of code from macros for better error reporting, and it's probably essentially an oversight that they appear in the final output at all.  I suppose they allow a reader to see the origin of code too but that seems quite a stretch, and `swig -E` would still show the locators if someone really wanted to see.

**Comment by wsfulton** [comment]
Another way to remove these markers is not to use the ruby_aux_method macro in rubyprimtypes.swg. I reckon I can remove it by a rewrite. There are lots of preprocessor hacks, hopefully I can figure one out.

**Comment by ojwb** [comment]
That would only solve this for SWIG/Ruby - the problem affects at least SWIG/Perl and SWIG/PHP too:

```
$ for d in Examples/*/{constants,simple} ; do make -C "$d" ; done
```
[snip]
```
$ grep '@SWIG:' Examples/*/{constants,simple}/*_wrap.c*
Examples/perl5/constants/example_wrap.c:  /*@SWIG:../../../Lib/perl5/perltypemaps.swg,67,%set_constant@*/ do {
Examples/perl5/constants/example_wrap.c:  /*@SWIG:../../../Lib/perl5/perltypemaps.swg,67,%set_constant@*/ do {
Examples/perl5/constants/example_wrap.c:  /*@SWIG:../../../Lib/perl5/perltypemaps.swg,67,%set_constant@*/ do {
Examples/perl5/constants/example_wrap.c:  /*@SWIG:../../../Lib/perl5/perltypemaps.swg,67,%set_constant@*/ do {
Examples/perl5/constants/example_wrap.c:  /*@SWIG:../../../Lib/perl5/perltypemaps.swg,67,%set_constant@*/ do {
Examples/perl5/constants/example_wrap.c:  /*@SWIG:../../../Lib/perl5/perltypemaps.swg,67,%set_constant@*/ do {
Examples/perl5/constants/example_wrap.c:  /*@SWIG:../../../Lib/perl5/perltypemaps.swg,67,%set_constant@*/ do {
Examples/perl5/constants/example_wrap.c:  /*@SWIG:../../../Lib/perl5/perltypemaps.swg,67,%set_constant@*/ do {
Examples/perl5/constants/example_wrap.c:  /*@SWIG:../../../Lib/perl5/perltypemaps.swg,67,%set_constant@*/ do {
Examples/php/simple/example_wrap.c:  /*@SWIG:../../../Lib/php/utils.i,6,CONVERT_INT_IN@*/
Examples/php/simple/example_wrap.c:  /*@SWIG:../../../Lib/php/utils.i,6,CONVERT_INT_IN@*/
Examples/php/simple/example_wrap.c:  /*@SWIG:../../../Lib/php/utils.i,48,CONVERT_FLOAT_IN@*/
Examples/ruby/simple/example_wrap.c:/*@SWIG:../../../Lib/ruby/rubyprimtypes.swg,19,%ruby_aux_method@*/
Examples/ruby/simple/example_wrap.c:/*@SWIG:../../../Lib/ruby/rubyprimtypes.swg,19,%ruby_aux_method@*/
```

It's possible other target languages are affected, but only in situations not exercised by either of these examples.

**Comment by robinst** [comment]
> but maybe better would be to work out how these get to the final generated output and stop that happening

I had a look but it seems to be strings all the way down, starting from the string returned from `expand_macro`, that gets called by `Preprocessor_replace` which produces a bigger string including that. And I don't know enough about the code to know where it would have to be re-parsed and stripped vs where it has to be included.

As for the suggestion in the [first reply](https://github.com/swig/swig/issues/202#issuecomment-245805763), I had a look too: `SwigLib` is defined in `main.cxx` and only used there, not sure how it could be accessed from `scanner.c`? Maybe it could be passed into the `Scanner`, or whatever reads the scanner locations could do the substitution.

**Comment by ojwb** [comment]
Yes, the representation is just text up until the lexer - that's why this location information is encoded as added magic comments.  Conceivably the location information could be stored in some sort of external data structure instead (e.g. byte span 42 to 99 came from file F lines M-N, and could even record the macro name for improved error reported).  That would be cleaner and should make this problem go away, but this external data structure would need updating if the text gets modified after that point (I'm not sure if it is or not).

These `@SWIG:`...`@` "locator strings" are parsed by `Scanner_locator()` in `Source/Swig/scanner.c`, but I think in this case this output probably isn't going through the lexer (or if it is, there's a separate copy).  I don't know exactly what's going on here though.

**Comment by ojwb** [comment]
> could even record the macro name for improved error reported

Actually the macro name is already in the magic comments.

**Comment by wsfulton** [comment]
2f194c10a8732c928082af795dbee5f7d7ad37e2 contains a tactical improvement by removing the vast majority of locator strings for Javascript, MzScheme, Python and Ruby by using fewer/simpler macros.

The test-suite for these 3 are much improved with locators now only coming from 
1. macros in the actual testcase in the case of primitive_types
2. macros in factory.i that defeated me: there are vararg macro which may not even be standard C preprocessing.

I'm not very familiar with the locator strings, exactly when they are needed and why they appear sometimes and other times not. However, while experimenting I noticed that they will disappear when a multi-line macro is replaced with a single line macro. That could solve a few more similar cases at the expense of less readable generated code.

The other target languages are almost perfectly clean apart from Perl and PHP where the test-suite is still full of these macros. Perl looks non-trivial to me, but PHP might be possible to improve by using duplicated snippets of code in typemaps instead of macros by removing the CONVERT macros. That's not for me to say though and one for @ojwb to call. I'm not doing any more tactical improvements in this area; understanding the locators fully and then not issuing them unnecessarily into the generated code would be the proper fix, but I don't plan on working on this.

**Comment by ojwb** [comment]
> I noticed that they will disappear when a multi-line macro is replaced with a single line macro.

That'll be because there's an explicit check for whether the expansion contains a `'\n'` and the locator comments are only added if it does.

I don't understand them in full, but one of the purposes seems to be to keep source line number counting correct when a macro expands to multiple lines.  A pure C/C++ preprocessor could perhaps solve that by replacing newlines with other whitespace in the expansion, but SWIG's macros could be used inside something like `%pythoncode` so a change like that might be undesirable).

I'm not very keen on the idea of replacing use of macros with a lot of repeated code as that seems to be the wrong direction to be going.  Changing the expansions to be a single line isn't ideal but seems a better option than not using macros in any of these situations.
